File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int l = 0;
        int h = height.length - 1;
        int left = height[l];
        int right = height[h];
        int sum = 0;
        while (l < h)
        {
            if (left <= right)
            {
                l++;
                left = Math.max(left, height[l]);
                sum += left - height[l];
            }
            else
            {
                h--;
                right = Math.max(right, height[h]);
                sum += right - height[h];
            }
        }
        return sum;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        maxltor = [0] * len(height)
        for i in range(1, len(height)):
            maxltor[i] = max(maxltor[i - 1], height[i - 1])
        maxrtol = [0] * len(height)
        for i in range(len(height) - 1 - 1, -1, -1):
            maxrtol[i] = max(maxrtol[i + 1], height[i + 1])
        res = 0
        for i in range(len(height)):
            water = min(maxltor[i], maxrtol[i]) - height[i]
            if water > 0:
                res += water
        return res